File: server/projects.py
"""Project and scan management.

Handles project CRUD, scan creation, failure status updates.
All data persisted as JSON files under DATA_DIR.
Uses an index file for fast scan listing.
"""

import json
import threading
import uuid
from datetime import datetime, timezone
from pathlib import Path

from server.scanner import scan_project

DATA_DIR = Path("data")

_lock = threading.Lock()


def _projects_path():
    return DATA_DIR / "projects.json"


def _scans_dir():
    d = DATA_DIR / "scans"
    d.mkdir(parents=True, exist_ok=True)
    return d


def _scan_path(scan_id):
    return _scans_dir() / f"{scan_id}.json"


def _index_path():
    return _scans_dir() / "index.json"


def _read_json(path):
    if not path.is_file():
        return None
    with open(path) as f:
        return json.load(f)


def _write_json(path, data):
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(".tmp")
    with open(tmp, "w") as f:
        json.dump(data, f, indent=2)
    tmp.rename(path)


# --- Projects ---


def list_projects():
    with _lock:
        projects = _read_json(_projects_path()) or []
        migrated = False
        for p in projects:
            if "profiles" not in p:
                p["profiles"] = list(DEFAULT_PROFILES)
                migrated = True
        if migrated:
            _write_json(_projects_path(), projects)
        return projects


def _default_profiles():
    from server.templates import get_template, template_to_profile

    t = get_template("paparazzi")
    if t:
        return [template_to_profile(t)]
    return [
        {
            "name": "Paparazzi",
            "failures_dir": "build/paparazzi/failures",
            "golden_patterns": ["src/test/snapshots/images/{name}.png"],
        }
    ]


DEFAULT_PROFILES = _default_profiles()


def add_project(name, path, template_ids=None):
    from server import templates as tmpl

    if template_ids:
        profiles = []
        for tid in template_ids:
            t = tmpl.get_template(tid)
            if t:
                profiles.append(tmpl.template_to_profile(t))
    else:
        profiles = list(DEFAULT_PROFILES)

    with _lock:
        projects = _read_json(_projects_path()) or []
        project = {
            "id": uuid.uuid4().hex[:8],
            "name": name,
            "path": path,
            "added": datetime.now(timezone.utc).isoformat(),
            "profiles": profiles,
        }
        projects.append(project)
        _write_json(_projects_path(), projects)
    return project


def get_project(project_id):
    for p in list_projects():
        if p["id"] == project_id:
            return p
    return None


def update_project_profiles(project_id, profiles):
    with _lock:
        projects = _read_json(_projects_path()) or []
        for p in projects:
            if p["id"] == project_id:
                p["profiles"] = profiles
                break
        _write_json(_projects_path(), projects)
    return get_project(project_id)


def delete_project(project_id):
    with _lock:
        projects = _read_json(_projects_path()) or []
        projects = [p for p in projects if p["id"] != project_id]
        _write_json(_projects_path(), projects)


# --- Scan Index ---


def _scan_summary(scan):
    """Extract lightweight summary from a full scan for the index."""
    return {
        "id": scan["id"],
        "projectId": scan["projectId"],
        "projectName": scan.get("projectName", ""),
        "created": scan["created"],
        "modules": scan["modules"],
        "stats": scan["stats"],
    }


def _read_index():
    """Read the scan index, rebuilding from scan files if missing."""
    idx = _read_json(_index_path())
    if idx is not None:
        return idx
    return _rebuild_index()


def _rebuild_index():
    """Rebuild index from existing scan files. Called once on migration."""
    index = []
    scans_dir = _scans_dir()
    for f in sorted(scans_dir.glob("*.json"), reverse=True):
        if f.name == "index.json":
            continue
        data = _read_json(f)
        if data:
            index.append(_scan_summary(data))
    _write_json(_index_path(), index)
    return index


def _update_index(summary):
    """Add a scan summary to the index."""
    index = _read_index()
    index.insert(0, summary)
    _write_json(_index_path(), index)


def _remove_from_index(scan_id):
    """Remove a scan from the index."""
    index = _read_index()
    index = [s for s in index if s["id"] != scan_id]
    _write_json(_index_path(), index)


# --- Scans ---


def create_scan(project_id):
    """Blocking scan — used by tests and backward compat."""
    project = get_project(project_id)
    if not project:
        return None

    result = scan_project(project["path"])
    return create_scan_from_results(
        project_id,
        project["name"],
        project["path"],
        result["modules"],
        result["failures"],
    )


def create_scan_from_results(project_id, project_name, project_path, modules, failures):
    """Create and persist a scan from pre-computed results. Called by scan_jobs.

    Enforces one scan per project: deletes any existing scans for this project
    before creating the new one (atomically under the lock).
    """
    scan_id = datetime.now().strftime("%Y%m%d-%H%M%S") + "-" + uuid.uuid4().hex[:4]
    scan = {
        "id": scan_id,
        "projectId": project_id,
        "projectName": project_name,
        "projectPath": project_path,
        "created": datetime.now(timezone.utc).isoformat(),
        "modules": modules,
        "failures": failures,
        "stats": _compute_stats(failures),
    }

    with _lock:
        # One scan per project: remove old scans atomically
        index = _read_index()
        old_ids = [s["id"] for s in index if s["projectId"] == project_id]
        for old_id in old_ids:
            p = _scan_path(old_id)
            if p.is_file():
                p.unlink()
        index = [s for s in index if s["id"] not in set(old_ids)]

        _write_json(_scan_path(scan_id), scan)
        index.insert(0, _scan_summary(scan))
        _write_json(_index_path(), index)

    return scan


def list_scans():
    with _lock:
        return _read_index()


def get_scan(
    scan_id,
    page=0,
    size=50,
    status=None,
    query=None,
    module=None,
    profile=None,
    sort=None,
):
    with _lock:
        scan = _read_json(_scan_path(scan_id))
    if not scan:
        return None

    failures = scan["failures"]

    if status and status != "all":
        failures = [f for f in failures if f["status"] == status]
    if module:
        failures = [f for f in failures if f["module"] == module]
    if profile:
        failures = [f for f in failures if f.get("profile") == profile]
    if query:
        q = query.lower()
        failures = [
            f
            for f in failures
            if q in f["filename"].lower()
            or q in f["package"].lower()
            or q in f["class_name"].lower()
            or q in f["method"].lower()
        ]

    # Sort
    if sort == "name":
        failures.sort(key=lambda f: f.get("filename", "").lower())
    elif sort == "module":
        failures.sort(
            key=lambda f: (f.get("module", ""), f.get("filename", "").lower())
        )
    elif sort == "profile":
        failures.sort(
            key=lambda f: (f.get("profile", ""), f.get("filename", "").lower())
        )
    elif sort == "diff":
        failures.sort(key=lambda f: f.get("diff_pct") or 0, reverse=True)

    total_filtered = len(failures)
    start = page * size
    page_failures = failures[start : start + size]

    return {
        "id": scan["id"],
        "projectId": scan["projectId"],
        "projectName": scan.get("projectName", ""),
        "projectPath": scan.get("projectPath", ""),
        "created": scan["created"],
        "modules": scan["modules"],
        "stats": scan["stats"],
        "failures": page_failures,
        "totalFiltered": total_filtered,
        "page": page,
        "pageSize": size,
    }


def delete_scan(scan_id):
    with _lock:
        path = _scan_path(scan_id)
        if path.is_file():
            path.unlink()
        _remove_from_index(scan_id)


def update_scan_module(scan_id, module_name, module_data, module_failures):
    """Update a single module's data in an existing scan (in-place)."""
    with _lock:
        scan = _read_json(_scan_path(scan_id))
        if not scan:
            return None

        # Replace or add module entry
        found = False
        for i, m in enumerate(scan["modules"]):
            if m["name"] == module_name:
                scan["modules"][i] = module_data
                found = True
                break
        if not found:
            scan["modules"].append(module_data)

        # Replace failures for this module
        scan["failures"] = [
            f for f in scan["failures"] if f["module"] != module_name
        ] + module_failures

        scan["stats"] = _compute_stats(scan["failures"])
        _write_json(_scan_path(scan_id), scan)

        # Update index
        idx = _read_index()
        for i, s in enumerate(idx):
            if s["id"] == scan_id:
                idx[i] = _scan_summary(scan)
                break
        _write_json(_index_path(), idx)

    return scan["stats"]


def update_failure_status(scan_id, filename, status):
    with _lock:
        scan = _read_json(_scan_path(scan_id))
        if not scan:
            return None
        for f in scan["failures"]:
            if f["filename"] == filename:
                f["status"] = status
                break
        scan["stats"] = _compute_stats(scan["failures"])
        _write_json(_scan_path(scan_id), scan)
    return scan["stats"]


def batch_update_status(scan_id, filenames, status):
    with _lock:
        scan = _read_json(_scan_path(scan_id))
        if not scan:
            return None
        target = set(filenames)
        for f in scan["failures"]:
            if f["filename"] in target:
                f["status"] = status
        scan["stats"] = _compute_stats(scan["failures"])
        _write_json(_scan_path(scan_id), scan)
    return scan["stats"]


def is_path_under_project(file_path):
    """Validate that a file path is under a registered project directory."""
    projects = list_projects()
    resolved = Path(file_path).resolve()
    for p in projects:
        if resolved.is_relative_to(Path(p["path"]).resolve()):
            return True
    return False


def _compute_stats(failures):
    stats = {"total": len(failures), "pending": 0, "accepted": 0, "rejected": 0}
    for f in failures:
        s = f.get("status", "pending")
        if s in stats:
            stats[s] += 1
    return stats

File: src/projects.js
/**
 * Project and scan management.
 * Handles project CRUD, scan creation, failure status updates.
 * All data persisted as JSON files under DATA_DIR.
 * Uses an index file for fast scan listing.
 */

const fs = require('fs');
const os = require('os');
const path = require('path');
const crypto = require('crypto');
const templates = require('./templates');
const { readJson, writeJson } = require('./jsonStore');

function expandHome(p) {
  if (typeof p !== 'string' || !p) return p;
  if (p === '~') return os.homedir();
  if (p.startsWith('~/') || p.startsWith('~\\')) return path.join(os.homedir(), p.slice(2));
  return p;
}

let DATA_DIR = path.join(process.cwd(), 'data');

function setDataDir(dir) {
  DATA_DIR = dir;
  _scansDirCreated = false;
  templates.setDataDir(dir);
}

function projectsPath() { return path.join(DATA_DIR, 'projects.json'); }
let _scansDirCreated = false;
function scansDir() {
  const d = path.join(DATA_DIR, 'scans');
  if (!_scansDirCreated) {
    fs.mkdirSync(d, { recursive: true });
    _scansDirCreated = true;
  }
  return d;
}
function sanitizeId(id) {
  return String(id).replace(/[^a-zA-Z0-9._-]/g, '');
}

function scanPath(scanId) { return path.join(scansDir(), `${sanitizeId(scanId)}.json`); }
function indexPath() { return path.join(scansDir(), 'index.json'); }

// --- Default profiles ---

let _defaultProfiles = null;

function defaultProfiles() {
  if (_defaultProfiles) return _defaultProfiles;
  const t = templates.getTemplate('paparazzi');
  if (t) {
    _defaultProfiles = [templates.templateToProfile(t)];
  } else {
    _defaultProfiles = [{
      name: 'Paparazzi',
      failures_dir: 'build/paparazzi/failures',
      golden_patterns: ['src/test/snapshots/images/{name}.png'],
    }];
  }
  return _defaultProfiles;
}

// --- Projects ---

function listProjects() {
  const list = readJson(projectsPath()) || [];
  return list.map(p => (p && typeof p.path === 'string' ? { ...p, path: expandHome(p.path) } : p));
}

function addProject(name, projectPath, templateIds, strategy) {
  projectPath = expandHome(projectPath);
  let profiles;
  if (templateIds && templateIds.length) {
    profiles = [];
    for (const tid of templateIds) {
      const t = templates.getTemplate(tid);
      if (t) profiles.push(templates.templateToProfile(t));
    }
  } else {
    profiles = JSON.parse(JSON.stringify(defaultProfiles()));
  }

  const projects = readJson(projectsPath()) || [];
  const project = {
    id: crypto.randomBytes(4).toString('hex'),
    name,
    path: projectPath,
    added: new Date().toISOString(),
    strategy: strategy || 'gradle',
    profiles,
  };
  projects.push(project);
  writeJson(projectsPath(), projects);
  return project;
}

function getProject(projectId) {
  return listProjects().find(p => p.id === projectId) || null;
}

function updateProjectProfiles(projectId, profiles) {
  const projects = readJson(projectsPath()) || [];
  for (const p of projects) {
    if (p.id === projectId) {
      p.profiles = profiles;
      break;
    }
  }
  writeJson(projectsPath(), projects);
  return getProject(projectId);
}

function deleteProject(projectId) {
  const projects = readJson(projectsPath()) || [];
  writeJson(projectsPath(), projects.filter(p => p.id !== projectId));
}

// --- Scan Index ---

function scanSummary(scan) {
  return {
    id: scan.id,
    projectId: scan.projectId,
    projectName: scan.projectName || '',
    created: scan.created,
    modules: scan.modules,
    stats: scan.stats,
  };
}

function readIndex() {
  const idx = readJson(indexPath());
  if (idx !== null) return idx;
  return rebuildIndex();
}

function rebuildIndex() {
  const dir = scansDir();
  const files = fs.readdirSync(dir).filter(f => f.endsWith('.json') && f !== 'index.json').sort().reverse();
  const index = [];
  for (const f of files) {
    const data = readJson(path.join(dir, f));
    if (data) index.push(scanSummary(data));
  }
  writeJson(indexPath(), index);
  return index;
}

function updateIndex(summary) {
  const index = readIndex();
  index.unshift(summary);
  writeJson(indexPath(), index);
}

function removeFromIndex(scanId) {
  const index = readIndex();
  writeJson(indexPath(), index.filter(s => s.id !== scanId));
}

// --- Scans ---

function createScanFromResults(projectId, projectName, projectPath, modules, failures) {
  const now = new Date();
  const ts = now.toISOString().replace(/[-:T]/g, '').slice(0, 14).replace(/^(\d{8})(\d{6})/, '$1-$2');
  const scanId = ts + '-' + crypto.randomBytes(2).toString('hex');
  const scan = {
    id: scanId,
    projectId,
    projectName,
    projectPath,
    created: now.toISOString(),
    modules,
    failures,
    stats: computeStats(failures),
  };

  // One scan per project: remove old scans atomically
  const index = readIndex();
  const oldIds = index.filter(s => s.projectId === projectId).map(s => s.id);
  for (const oldId of oldIds) {
    const p = scanPath(oldId);
    if (fs.existsSync(p)) fs.unlinkSync(p);
  }
  const cleanIndex = index.filter(s => !oldIds.includes(s.id));

  writeJson(scanPath(scanId), scan);
  cleanIndex.unshift(scanSummary(scan));
  writeJson(indexPath(), cleanIndex);

  return scan;
}

function listScans() {
  return readIndex();
}

function getScan(scanId, { page = 0, size = 50, status, query, module: mod, profile, sort } = {}) {
  const scan = readJson(scanPath(scanId));
  if (!scan) return null;

  let failures = scan.failures;

  if (status && status !== 'all') {
    failures = failures.filter(f => f.status === status);
  }
  if (mod) {
    failures = failures.filter(f => f.module === mod);
  }
  if (profile) {
    failures = failures.filter(f => f.profile === profile);
  }
  if (query) {
    const q = query.toLowerCase();
    failures = failures.filter(f =>
      f.filename.toLowerCase().includes(q) ||
      f.package.toLowerCase().includes(q) ||
      f.class_name.toLowerCase().includes(q) ||
      f.method.toLowerCase().includes(q)
    );
  }

  if (sort === 'name') {
    failures.sort((a, b) => (a.filename || '').toLowerCase().localeCompare((b.filename || '').toLowerCase()));
  } else if (sort === 'module') {
    failures.sort((a, b) => (a.module || '').localeCompare(b.module || '') || (a.filename || '').toLowerCase().localeCompare((b.filename || '').toLowerCase()));
  } else if (sort === 'profile') {
    failures.sort((a, b) => (a.profile || '').localeCompare(b.profile || '') || (a.filename || '').toLowerCase().localeCompare((b.filename || '').toLowerCase()));
  } else if (sort === 'diff') {
    failures.sort((a, b) => (b.diff_pct || 0) - (a.diff_pct || 0));
  }

  const totalFiltered = failures.length;
  const start = page * size;
  const pageFailures = failures.slice(start, start + size);

  return {
    id: scan.id,
    projectId: scan.projectId,
    projectName: scan.projectName || '',
    projectPath: scan.projectPath || '',
    created: scan.created,
    modules: scan.modules,
    stats: scan.stats,
    failures: pageFailures,
    totalFiltered,
    page,
    pageSize: size,
  };
}

function deleteScan(scanId) {
  const p = scanPath(scanId);
  if (fs.existsSync(p)) fs.unlinkSync(p);
  removeFromIndex(scanId);
}

function updateScanModule(scanId, moduleName, moduleData, moduleFailures) {
  const scan = readJson(scanPath(scanId));
  if (!scan) return null;

  let found = false;
  for (let i = 0; i < scan.modules.length; i++) {
    if (scan.modules[i].name === moduleName) {
      scan.modules[i] = moduleData;
      found = true;
      break;
    }
  }
  if (!found) scan.modules.push(moduleData);

  // Preserve accept/reject decisions across rescans: a re-detected failure with
  // the same filename keeps the user's prior status until it actually goes away.
  const priorStatus = new Map();
  for (const f of scan.failures) {
    if (f.module === moduleName && (f.status === 'accepted' || f.status === 'rejected')) {
      priorStatus.set(f.filename, f.status);
    }
  }
  for (const f of moduleFailures) {
    const prior = priorStatus.get(f.filename);
    if (prior) f.status = prior;
  }

  scan.failures = scan.failures.filter(f => f.module !== moduleName).concat(moduleFailures);
  scan.stats = computeStats(scan.failures);
  writeJson(scanPath(scanId), scan);

  const idx = readIndex();
  for (let i = 0; i < idx.length; i++) {
    if (idx[i].id === scanId) {
      idx[i] = scanSummary(scan);
      break;
    }
  }
  writeJson(indexPath(), idx);

  return scan.stats;
}

function updateFailureStatus(scanId, filename, status) {
  const scan = readJson(scanPath(scanId));
  if (!scan) return null;
  for (const f of scan.failures) {
    if (f.filename === filename) {
      f.status = status;
      break;
    }
  }
  scan.stats = computeStats(scan.failures);
  writeJson(scanPath(scanId), scan);
  return scan.stats;
}

function batchUpdateStatus(scanId, filenames, status) {
  const scan = readJson(scanPath(scanId));
  if (!scan) return null;
  const target = new Set(filenames);
  for (const f of scan.failures) {
    if (target.has(f.filename)) f.status = status;
  }
  scan.stats = computeStats(scan.failures);
  writeJson(scanPath(scanId), scan);
  return scan.stats;
}

// Returns null on success, error string on failure.
function _copyActualToGolden(failure, rootPrefix) {
  if (!failure.actual_path) return 'no actual image to accept';
  if (!failure.golden_path) return 'no golden path resolved; check profile patterns';

  let actualReal;
  try { actualReal = fs.realpathSync(failure.actual_path); } catch { return 'actual image missing'; }

  // Create parent dir first so we can realpath it (golden file itself may not exist yet)
  let goldenResolved;
  try {
    fs.mkdirSync(path.dirname(failure.golden_path), { recursive: true });
    goldenResolved = path.join(fs.realpathSync(path.dirname(failure.golden_path)), path.basename(failure.golden_path));
  } catch (e) { return 'golden path invalid: ' + e.message; }

  if (!actualReal.startsWith(rootPrefix)) return 'actual outside project';
  if (!goldenResolved.startsWith(rootPrefix)) return 'golden outside project';

  try { fs.copyFileSync(actualReal, goldenResolved); }
  catch (e) { return 'copy failed: ' + e.message; }
  return null;
}

function _getScanAndRoot(scanId) {
  const scan = readJson(scanPath(scanId));
  if (!scan) return { scan: null };
  const project = getProject(scan.projectId);
  if (!project) return { scan, error: 'project not found' };
  let root;
  try { root = fs.realpathSync(project.path); } catch { return { scan, error: 'project path invalid' }; }
  return { scan, rootPrefix: root + path.sep };
}

function acceptBaseline(scanId, filename) {
  const { scan, rootPrefix, error } = _getScanAndRoot(scanId);
  if (!scan) return null;
  if (error) return { error };
  const failure = scan.failures.find(f => f.filename === filename);
  if (!failure) return null;

  const err = _copyActualToGolden(failure, rootPrefix);
  if (err) return { error: err };

  failure.status = 'accepted';
  scan.stats = computeStats(scan.failures);
  writeJson(scanPath(scanId), scan);
  return { stats: scan.stats };
}

function acceptAllBaselines(scanId) {
  const { scan, rootPrefix, error } = _getScanAndRoot(scanId);
  if (!scan) return null;
  if (error) return { error };

  let accepted = 0;
  const errors = [];
  for (const failure of scan.failures) {
    if (failure.status === 'accepted') continue;
    const err = _copyActualToGolden(failure, rootPrefix);
    if (err) {
      errors.push({ filename: failure.filename, error: err });
    } else {
      failure.status = 'accepted';
      accepted++;
    }
  }

  scan.stats = computeStats(scan.failures);
  writeJson(scanPath(scanId), scan);
  return { stats: scan.stats, accepted, errors };
}

function isPathUnderProject(filePath) {
  const allProjects = listProjects();
  const resolved = path.resolve(filePath);
  return allProjects.some(p => {
    const projectRoot = path.resolve(p.path) + path.sep;
    return resolved.startsWith(projectRoot) || resolved === path.resolve(p.path);
  });
}

function computeStats(failures) {
  const stats = { total: failures.length, pending: 0, accepted: 0, rejected: 0 };
  for (const f of failures) {
    const s = f.status || 'pending';
    if (s in stats) stats[s]++;
  }
  return stats;
}

function getDataDir() { return DATA_DIR; }

function importProjects(incoming) {
  const existing = readJson(projectsPath()) || [];
  const byId = new Map(existing.map(p => [p.id, p]));
  for (const p of incoming) {
    byId.set(p.id, p);
  }
  writeJson(projectsPath(), [...byId.values()]);
}

function resetProjects() {
  writeJson(projectsPath(), []);
}

module.exports = {
  expandHome,
  setDataDir,
  getDataDir,
  listProjects,
  addProject,
  getProject,
  updateProjectProfiles,
  deleteProject,
  importProjects,
  resetProjects,
  listScans,
  getScan,
  createScanFromResults,
  deleteScan,
  updateScanModule,
  updateFailureStatus,
  batchUpdateStatus,
  acceptBaseline,
  acceptAllBaselines,
  scanPath,
};
